src/WebView/PreviewPanel.ts

```diff
@@ -0,0 +1,191 @@
+import { Query } from '../ContentProvider/openHAB';
+
+/**
+ * Manages the extension WebView panel
+ * 
+ * Reference implementation accessable through
+ * https://github.com/Microsoft/vscode-extension-samples/tree/master/webview-sample
+ * 
+ * @author Jerome Luckenbach - Initial contribution
+ */
+export class PreviewPanel {
+
+     /**
+     * Track the currently panel. Only allow a single panel to exist at a time.
+     */
+    public static currentPanel: PreviewPanel | undefined;
+
+    public static readonly viewType = 'ohPreviewPanel';
+
+    private readonly _panel: vscode.WebviewPanel;
+    private readonly _extensionPath: string;
+    private _disposables: vscode.Disposable[] = [];
+
+    public static createOrShow(extensionPath: string, title : string, query : Query) {
+
+        // If we already have a panel, show it.
+        if (PreviewPanel.currentPanel) {
+            PreviewPanel.currentPanel._panel.reveal(vscode.ViewColumn.Two);
+            return;
+        }
+
+        // Otherwise, create a new panel.
+        const panel = vscode.window.createWebviewPanel(PreviewPanel.viewType, "openHAB Preview", vscode.ViewColumn.Two, {
+            // Enable javascript in the webview
+            enableScripts: true
+        });
+
+        PreviewPanel.currentPanel = new PreviewPanel(panel, extensionPath);
+    }
+
+    public static revive(panel: vscode.WebviewPanel, extensionPath: string) {
+        PreviewPanel.currentPanel = new PreviewPanel(panel, extensionPath);
+    }
+
+    private constructor(panel: vscode.WebviewPanel, extensionPath: string) {
+        this._panel = panel;
+        this._extensionPath = extensionPath;
+
+        // Set the webview's initial html content 
+        this._update();
+
+        // Listen for when the panel is disposed
+        // This happens when the user closes the panel or when the panel is closed programatically
+        this._panel.onDidDispose(() => this.dispose(), null, this._disposables);
+
+        // Update the content based on view changes
+        this._panel.onDidChangeViewState(e => {
+            if (this._panel.visible) {
+                this._update()
+            }
+        }, null, this._disposables);
+
+        // Handle messages from the webview
+        this._panel.webview.onDidReceiveMessage(message => {
+            switch (message.command) {
+                case 'alert':
+                    vscode.window.showErrorMessage(message.text);
+                    return;
+            }
+        }, null, this._disposables);
+    }
+
+    public doRefactor() {
+        // Send a message to the webview.
+        // You can send any JSON serializable data.
+        this._panel.webview.postMessage({ command: 'refactor' });
+    }
+
+    // Idea: doRefresh method here for auto updating on save
+
+    public dispose() {
+        PreviewPanel.currentPanel = undefined;
+
+        // Clean up our resources
+        this._panel.dispose();
+
+        while (this._disposables.length) {
```

Comment:
Instead of while loop with pop you could also do something like
```
for (const disposable of this._disposables) {
   disposable.dispose()
}
```
or alternatively
```
this._disposables.forEach((disposable)=>disposable.dispose())
```

Reply:
The code is mainly inspired/derived from https://github.com/Microsoft/vscode-extension-samples/tree/master/webview-sample i will leave it for now.

Thanks anyway.